File: frontend/src/lib/stores/onboarding.ts
import { writable } from 'svelte/store';
import { api, ApiError } from '$lib/api';

// Types
export interface Rule {
	id: string;
	order: number;
	title: string;
	description: string;
}

export interface ScreeningQuestion {
	id: string;
	order: number;
	question: string;
	required: boolean;
	type: 'text' | 'select' | 'agree';
	options?: string[];
}

export interface ScreeningAnswer {
	question_id: string;
	answer: string;
}

export interface WelcomeScreen {
	id: string;
	server_id: string;
	enabled: boolean;
	title: string;
	description: string;
	welcome_channels: string[];
	updated_at: string;
	created_at: string;
}

export interface WelcomeScreenConfig extends WelcomeScreen {
	rules: Rule[];
	questions: ScreeningQuestion[];
}

export interface MemberScreening {
	id: string;
	user_id: string;
	server_id: string;
	answers: ScreeningAnswer[];
	rules_read: boolean;
	status: 'pending' | 'approved' | 'rejected';
	created_at: string;
	updated_at: string;
}

// Stores
export const welcomeScreenConfig = writable<WelcomeScreenConfig | null>(null);
export const myScreening = writable<MemberScreening | null>(null);
export const pendingScreenings = writable<MemberScreening[]>([]);
export const welcomeLoading = writable(false);
export const welcomeError = writable<string | null>(null);

// API functions
export async function loadWelcomeScreen(serverId: string): Promise<WelcomeScreenConfig | null> {
	welcomeLoading.set(true);
	welcomeError.set(null);

	try {
		const config = await api.get<WelcomeScreenConfig>(`/servers/${serverId}/welcome`);
		welcomeScreenConfig.set(config);
		return config;
	} catch (error) {
		console.error('Failed to load welcome screen:', error);
		if (error instanceof ApiError && error.status !== 404) {
			welcomeError.set(error.message);
		}
		return null;
	} finally {
		welcomeLoading.set(false);
	}
}

export async function updateWelcomeScreen(
	serverId: string,
	updates: Partial<WelcomeScreenConfig>
): Promise<WelcomeScreenConfig> {
	welcomeLoading.set(true);
	welcomeError.set(null);

	try {
		const config = await api.put<WelcomeScreenConfig>(`/servers/${serverId}/welcome`, updates);
		welcomeScreenConfig.set(config);
		return config;
	} catch (error) {
		console.error('Failed to update welcome screen:', error);
		if (error instanceof ApiError) {
			welcomeError.set(error.message);
		}
		throw error;
	} finally {
		welcomeLoading.set(false);
	}
}

export async function submitScreening(
	serverId: string,
	answers: ScreeningAnswer[],
	rulesRead: boolean
): Promise<MemberScreening> {
	const screening = await api.post<MemberScreening>(`/servers/${serverId}/screening`, {
		answers,
		rules_read: rulesRead,
	});
	myScreening.set(screening);
	return screening;
}

export async function getMyScreening(serverId: string): Promise<MemberScreening | null> {
	try {
		const screening = await api.get<MemberScreening>(`/servers/${serverId}/screening/me`);
		myScreening.set(screening);
		return screening;
	} catch (error) {
		if (error instanceof ApiError && error.status === 404) {
			myScreening.set(null);
			return null;
		}
		throw error;
	}
}

export async function loadPendingScreenings(
	serverId: string,
	limit = 50,
	offset = 0
): Promise<MemberScreening[]> {
	welcomeLoading.set(true);
	welcomeError.set(null);

	try {
		const screenings = await api.get<MemberScreening[]>(
			`/servers/${serverId}/screening/pending?limit=${limit}&offset=${offset}`
		);
		pendingScreenings.set(screenings);
		return screenings;
	} catch (error) {
		console.error('Failed to load pending screenings:', error);
		if (error instanceof ApiError) {
			welcomeError.set(error.message);
		}
		throw error;
	} finally {
		welcomeLoading.set(false);
	}
}

export async function approveScreening(serverId: string, userId: string): Promise<void> {
	await api.post(`/servers/${serverId}/screening/${userId}/approve`);
	pendingScreenings.update((s) => s.filter((screen) => screen.user_id !== userId));
}

export async function rejectScreening(serverId: string, userId: string, reason?: string): Promise<void> {
	await api.post(`/servers/${serverId}/screening/${userId}/reject`, { reason });
	pendingScreenings.update((s) => s.filter((screen) => screen.user_id !== userId));
}

File: backend/internal/api/handlers/onboarding.go
package handlers

import (
	"strconv"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"

	"hearth/internal/models"
	"hearth/internal/services"
)

// WelcomeHandler handles welcome screen HTTP endpoints
type WelcomeHandler struct {
	welcomeService *services.WelcomeService
	userService    UserServiceInterface
}

// NewWelcomeHandler creates a new welcome handler
func NewWelcomeHandler(welcomeService *services.WelcomeService, userService UserServiceInterface) *WelcomeHandler {
	return &WelcomeHandler{
		welcomeService: welcomeService,
		userService:    userService,
	}
}

// GetWelcomeScreen returns the welcome screen configuration for a server
// @Summary Get welcome screen
// @Description Returns the welcome screen configuration including rules and screening questions
// @Tags Servers
// @Produce json
// @Param id path string true "Server ID"
// @Success 200 {object} models.WelcomeScreenConfig
// @Failure 400 {object} fiber.Map "Invalid server ID"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/welcome [get]
func (h *WelcomeHandler) GetWelcomeScreen(c *fiber.Ctx) error {
	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	config, err := h.welcomeService.GetWelcomeScreen(c.Context(), serverID)
	if err != nil {
		return HandleServiceError(c, err)
	}

	return c.JSON(config)
}

// UpdateWelcomeScreen updates the welcome screen configuration
// @Summary Update welcome screen
// @Description Updates the welcome screen configuration including rules and screening questions
// @Tags Servers
// @Accept json
// @Produce json
// @Param id path string true "Server ID"
// @Param body body models.UpdateWelcomeScreenRequest true "Welcome screen update data"
// @Success 200 {object} models.WelcomeScreenConfig
// @Failure 400 {object} fiber.Map "Invalid server ID or request body"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 403 {object} fiber.Map "Insufficient permissions"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/welcome [put]
func (h *WelcomeHandler) UpdateWelcomeScreen(c *fiber.Ctx) error {
	userID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	var req models.UpdateWelcomeScreenRequest
	if err := c.BodyParser(&req); err != nil {
		return ParseError(c, err)
	}

	config, err := h.welcomeService.UpdateWelcomeScreen(c.Context(), serverID, userID, &req)
	if err != nil {
		return HandleServiceError(c, err)
	}

	return c.JSON(config)
}

// SubmitScreening submits a member's screening answers
// @Summary Submit screening
// @Description New member submits screening answers for server approval
// @Tags Servers
// @Accept json
// @Produce json
// @Param id path string true "Server ID"
// @Param body body models.SubmitScreeningRequest true "Screening answers"
// @Success 200 {object} models.MemberScreening
// @Failure 400 {object} fiber.Map "Invalid server ID or request body"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 409 {object} fiber.Map "Screening already submitted"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/screening [post]
func (h *WelcomeHandler) SubmitScreening(c *fiber.Ctx) error {
	userID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	var req models.SubmitScreeningRequest
	if err := c.BodyParser(&req); err != nil {
		return ParseError(c, err)
	}

	screening, err := h.welcomeService.SubmitScreening(c.Context(), userID, serverID, &req)
	if err != nil {
		return HandleServiceError(c, err)
	}

	return c.JSON(screening)
}

// GetMemberScreening returns the current user's screening status for a server
// @Summary Get my screening status
// @Description Returns the current user's screening status for a server
// @Tags Servers
// @Produce json
// @Param id path string true "Server ID"
// @Success 200 {object} models.MemberScreening
// @Failure 400 {object} fiber.Map "Invalid server ID"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 404 {object} fiber.Map "No screening found"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/screening/me [get]
func (h *WelcomeHandler) GetMemberScreening(c *fiber.Ctx) error {
	userID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	screening, err := h.welcomeService.GetMemberScreening(c.Context(), userID, serverID)
	if err != nil {
		return HandleServiceError(c, err)
	}
	if screening == nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"error": "no screening found",
		})
	}

	return c.JSON(screening)
}

// GetPendingScreenings returns pending screenings for a server (moderators only)
// @Summary Get pending screenings
// @Description Returns a list of pending member screenings for a server
// @Tags Servers
// @Produce json
// @Param id path string true "Server ID"
// @Param limit query int false "Number of screenings to return (default 50, max 100)"
// @Param offset query int false "Offset for pagination (default 0)"
// @Success 200 {array} models.MemberScreening
// @Failure 400 {object} fiber.Map "Invalid server ID"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 403 {object} fiber.Map "Insufficient permissions"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/screening/pending [get]
func (h *WelcomeHandler) GetPendingScreenings(c *fiber.Ctx) error {
	userID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	// Parse pagination
	limit, _ := strconv.Atoi(c.Query("limit", "50"))
	if limit <= 0 || limit > 100 {
		limit = 50
	}
	offset, _ := strconv.Atoi(c.Query("offset", "0"))
	if offset < 0 {
		offset = 0
	}

	screenings, err := h.welcomeService.GetPendingScreenings(c.Context(), serverID, userID, limit, offset)
	if err != nil {
		return HandleServiceError(c, err)
	}

	if screenings == nil {
		screenings = []*models.MemberScreening{}
	}

	return c.JSON(screenings)
}

// ScreeningDecisionRequest is the input for screening decisions
type ScreeningDecisionRequest struct {
	Reason string `json:"reason,omitempty"`
}

// ApproveScreening approves a member's screening (moderators only)
// @Summary Approve screening
// @Description Approves a member's screening application
// @Tags Servers
// @Param id path string true "Server ID"
// @Param userId path string true "User ID whose screening to approve"
// @Success 204 "Screening approved successfully"
// @Failure 400 {object} fiber.Map "Invalid server ID or user ID"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 403 {object} fiber.Map "Insufficient permissions"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/screening/{userId}/approve [post]
func (h *WelcomeHandler) ApproveScreening(c *fiber.Ctx) error {
	moderatorID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	userID, err := uuid.Parse(c.Params("userId"))
	if err != nil {
		return InvalidUUID(c, "user id")
	}

	if err := h.welcomeService.ApproveScreening(c.Context(), userID, serverID, moderatorID); err != nil {
		return HandleServiceError(c, err)
	}

	return c.SendStatus(fiber.StatusNoContent)
}

// RejectScreening rejects a member's screening (moderators only)
// @Summary Reject screening
// @Description Rejects a member's screening and removes them from the server
// @Tags Servers
// @Accept json
// @Param id path string true "Server ID"
// @Param userId path string true "User ID whose screening to reject"
// @Param body body ScreeningDecisionRequest false "Rejection reason"
// @Success 204 "Screening rejected successfully"
// @Failure 400 {object} fiber.Map "Invalid server ID or user ID"
// @Failure 401 {object} fiber.Map "Unauthorized"
// @Failure 403 {object} fiber.Map "Insufficient permissions"
// @Failure 500 {object} fiber.Map "Internal server error"
// @Router /servers/{id}/screening/{userId}/reject [post]
func (h *WelcomeHandler) RejectScreening(c *fiber.Ctx) error {
	moderatorID, err := getUserIDFromContext(c)
	if err != nil {
		return Unauthorized(c, "unauthorized")
	}

	serverID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return InvalidUUID(c, "server id")
	}

	userID, err := uuid.Parse(c.Params("userId"))
	if err != nil {
		return InvalidUUID(c, "user id")
	}

	var req ScreeningDecisionRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "invalid request body",
		})
	}

	if err := h.welcomeService.RejectScreening(c.Context(), userID, serverID, moderatorID, req.Reason); err != nil {
		return HandleServiceError(c, err)
	}

	return c.SendStatus(fiber.StatusNoContent)
}
